pkg/system_config/types.go

```diff
@@ -0,0 +1,228 @@
+package system_config
+
+import (
+	"fmt"
+	"github.com/BurntSushi/toml"
+	"k8s.io/apimachinery/pkg/util/json"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+const (
+	RegistriesConfPath = "/tmp/containers/registries.conf"
+	PolicyConfPath     = "/tmp/containers/policy.json"
+	DockerCertsDir     = "/tmp/docker/certs.d"
+	RegistryCertsDir   = "/etc/containers/registries.d"
+)
+
+type registryCertTuple struct {
+	registry string
+	cert     string
+}
+
+func (t registryCertTuple) writeToFile() error {
+	// create folder if it doesn't exist
+	absoluteFolderPath := fmt.Sprintf("%s/%s", DockerCertsDir, t.getFolderName())
+	if _, err := os.Stat(absoluteFolderPath); os.IsNotExist(err) {
+		err = os.MkdirAll(absoluteFolderPath, 0755)
+		if err != nil {
+			return err
+		}
+	}
+	// write cert to file
+	absoluteFilePath := fmt.Sprintf("%s/%s/ca.crt", DockerCertsDir, t.getFolderName())
+	f, err := os.Create(absoluteFilePath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	_, err = f.WriteString(t.cert)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (t registryCertTuple) getFolderName() string {
+	// the registry name could report the port number after two dots, e.g. registry.example.com..5000.
+	// we need to replace the two dots with a colon to get the correct folder name.
+	return strings.Replace(t.registry, "..", ":", 1)
+}
+
+type registriesConf struct {
+	UnqualifiedSearchRegistries []string                 `toml:"unqualified-search-registries"`
+	ShortNameMode               string                   `toml:"short-name-mode"`
+	Registries                  []*registryConf          `toml:"registries"`
+	registriesMap               map[string]*registryConf `toml:"-"`
+}
+
+func (rsc registriesConf) getRegistryConfOrCreate(registry string) *registryConf {
+	rc, _ := rsc.registriesMap[registry]
+	if rc == nil {
+		rc = &registryConf{
+			Location: registry,
+		}
+		rsc.registriesMap[registry] = rc
+		rsc.Registries = append(rsc.Registries, rc)
+	}
+	return rc
+}
+
+func (rsc registriesConf) writeToFile() error {
+	return writeTomlFile(RegistriesConfPath, rsc)
+}
+
+func (rsc registriesConf) getRegistryConf(registry string) (*registryConf, bool) {
+	rc, ok := rsc.registriesMap[registry]
+	return rc, ok
+}
+
+type registryConf struct {
+	Location string   `toml:"location"`
+	Prefix   string   `toml:"prefix"`
+	Mirrors  []string `toml:"mirror"`
+	// Setting the blocked, allowed and insecure fields to nil will cause them to be omitted from the output
+	Blocked  *bool `toml:"blocked"`
+	Allowed  *bool `toml:"allowed"`
+	Insecure *bool `toml:"insecure"`
+}
+
+// defaultRegistriesConf returns a default registriesConf object
+func defaultRegistriesConf() registriesConf {
+	return registriesConf{
+		UnqualifiedSearchRegistries: []string{"registry.access.redhat.com", "docker.io"},
```

Comment:
this is present in the node itself for all scenarios?

Reply:
yep.